### PR TITLE
UniPC Sampler support

### PR DESCRIPTION
**Pull request by Schorny**
this is more a discussion than a finalized PR

1) UniPC is the first sampler that is implemented locally in sdkit, not just imported. so we need a good location and naming guideline.

2) UniPC is also highly customizable with a few settings. how to expose them to the api?

3) we need good sample settings. i did a testrun with random settings and figured that
lower_order_final should always be True
thresholding should always be False

order=3 seems to be more vibrant sometimes
vary_coeff seems to create more detailed backgrounds

I created my sample settings on a quick look through my generated sample images. there is definitely way more to be done here.

### CODE REVIEW

**Comment by Schorny** [comment]
useful links:

automatic1111 feature request:
https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/7705
automatic1111 implementation (WIP):
https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/7710
official repo:
https://github.com/wl-zhao/UniPC

File: sdkit/generate/image_generator.py [hunk 1]
```diff
@@ -34,6 +34,7 @@ def generate_images(
         hypernetwork_strength: float = 0,
 
         callback=None,
+        sampler_params={}
```
Comment (cmdr2) [inline sdkit/generate/image_generator.py:37]: Please add a trailing comma (,)

File: sdkit/generate/sampler/unipc_samplers.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+#TODO: find best suggestions for sample params
+samplers = {
+    'unipc': { #suggested version
+        'variant': 'bh1',
+        'time_skip': 'logSNR',
+        'order': 3,
+        'lower_order_final': True,
+        'thresholding': False,
+    },
+    'unipc_snr': { #same as suggested version
```
Comment (cmdr2) [inline sdkit/generate/sampler/unipc_samplers.py:25]: Do we need duplicate entries? Maybe we can keep a single entry here (the snr one), and update the documentation to recommend one?

**Comment by cmdr2** [comment]
Hi @Schorny , thanks for this PR! Nice work!

1. Instead of `custom`, I'd suggest this folder structure:
```
sampler/
    sampler_main.py
    ...
    unipc_samplers/
        __init__.py # contains the code currently in unipc_samplers.py (with an 's')
        unipc_sampler.py # with the actual implementation
```

2. Regarding parameters, your `sampler_params` approach seems like the correct one. It allows us to keep sampler-related params in a single container, instead of flattening it into the function call.

Please ensure (if you haven't already) that good defaults are used, if no sampler params are passed.

Once this PR is merged, we can look at the Easy Diffusion UI separately.

Thanks again!

**Comment by cmdr2** [comment]
Hi @Schorny Thanks for this. IMO, the PR is only pending the folder structure change. No hurry or pressure, just trying to avoid any miscommunication.

**Comment by cmdr2** [comment]
Merged this to a branch, will test it out, thanks!

**Pull request merged**